### PR TITLE
[FIXED JENKINS-33520] make ATH work on windows and other issues.

### PR DESCRIPTION
**Pull request by jtnord**
ATH was just so badly broken on windows :cry: 

[JENKINS-33520](https://issues.jenkins-ci.org/browse/JENKINS-33520)

I beleive I have fixed most of the issues.
There is an open issue of how to detect the OS if running Jenkins remotely (however the current situation assumes it is run on the same OS type with the same tools (@native looks on the system running the ATH not the Jenkins system) - so this is a longer issue - and would significantly add to the test run time for the case where the OS does not match) - so is considered out of scope.

Many tests are skipped as Jenkins will not restart itself when started locally on Windows.  LocalController could possibly be enhanced to do a "manual" restart by stopping Jenkins and then starting it again.

Current state: 

- [x] Maven seem like genuine test failures (Maven has not been correctly added to the path as it is trying to execute mvn.bat and failing) - ([JENKINS-33693](https://issues.jenkins-ci.org/browse/JENKINS-33693))
- [x] some tests are still trying to execute bash scripts on windows :-/
- [ ] ~~I seem to have broken docker portmap (Failed to figure out port map 80)~~ `-DreuseForks=false` fixes this.  Something to investigate later.
- [ ] Firefox driver is unreliable (upgraded driver but still has issues)
- [x] Incorrect assumptions Docker
- [x] Incorrect assumptions in Job page object
- [x] @Native should not require cygwin
- [x] CIFSPublishPluginTest
- [x] PublishOverSSHPluginTest
- [x] CompressArtifactsPluginTest
- [x] DeployPluginTest
- [x] JacocoPluginTest - (fixed by [JENKINS-33693](https://issues.jenkins-ci.org/browse/JENKINS-33693))
- [x] JavadocPluginTest (fixed by [JENKINS-33693](https://issues.jenkins-ci.org/browse/JENKINS-33693))
- [x] LdapPluginTest
- [x] MavenPluginTest
- [x] CopyArtifactPluginTest (fixed by [JENKINS-33693](https://issues.jenkins-ci.org/browse/JENKINS-33693))
- [ ] ~~NodeLabelParameterPluginTest~~  not an issue in windows - tests are genuinley failing on Linux too ([JENKINS-33746](https://issues.jenkins-ci.org/browse/JENKINS-33746)).
- [x] Subversion_Version154_PluginTest
- [x] TaskScannerPluginTest (fixed by [JENKINS-33693](https://issues.jenkins-ci.org/browse/JENKINS-33693))
- [x] JiraContainer was hard coded to localhost
- [x] [JENKINS-33693](https://issues.jenkins-ci.org/browse/JENKINS-33693) 
- [ ] Random test failures with Guice cleanup errors :cry:
jenkinsci/jenkins/pull/2148
```
java.lang.AssertionError: org.jenkinsci.test.acceptance.FallbackConfig$1@4393593c failed
        at org.jenkinsci.test.acceptance.guice.Cleaner.performCleanUp(Cleaner.java:56)
        at org.jenkinsci.test.acceptance.guice.TestCleaner.performCleanUp(TestCleaner.java:19)
        at org.jenkinsci.test.acceptance.guice.World.endTestScope(World.java:66)
        at org.jenkinsci.test.acceptance.junit.JenkinsAcceptanceTestRule$1.evaluate(JenkinsAcceptanceTestRule.java:72)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
Caused by: org.openqa.selenium.remote.UnreachableBrowserException: Error communicating with the remote browser. It may have died.
Build info: version: '2.52.0', revision: '4c2593cfc3689a7fcd7be52549167e5ccc93ad28', time: '2016-02-11 11:22:43'
System info: host: 'cb-james', ip: '192.168.137.1', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_71'
Driver info: driver.version: EventFiringWebDriver
Capabilities [{applicationCacheEnabled=true, rotatable=false, handlesAlerts=true, databaseEnabled=true, version=45.0, platform=WINDOWS, nativeEvents=false, acceptSslCerts=true, webStorageEnabled=true, locationContextEnabled=true, browserName=firefox, takesScreenshot=true, javascriptEnabled=true, cssSelectorsEnabled=true}]
Session ID: 7d1711cd-d4c4-4130-bd75-94ac2b050851
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:665)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:701)
        at org.openqa.selenium.remote.RemoteWebDriver.quit(RemoteWebDriver.java:526)
        at sun.reflect.GeneratedMethodAccessor23.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.openqa.selenium.support.events.EventFiringWebDriver$2.invoke(EventFiringWebDriver.java:103)
        at com.sun.proxy.$Proxy30.quit(Unknown Source)
        at org.openqa.selenium.support.events.EventFiringWebDriver.quit(EventFiringWebDriver.java:202)
        at org.jenkinsci.test.acceptance.FallbackConfig$1.evaluate(FallbackConfig.java:168)
        at org.jenkinsci.test.acceptance.guice.Cleaner.performCleanUp(Cleaner.java:54)
        at org.jenkinsci.test.acceptance.guice.TestCleaner.performCleanUp(TestCleaner.java:19)
        at org.jenkinsci.test.acceptance.guice.World.endTestScope(World.java:66)
        at org.jenkinsci.test.acceptance.junit.JenkinsAcceptanceTestRule$1.evaluate(JenkinsAcceptanceTestRule.java:72)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
Caused by: java.net.SocketException: Permission denied: connect
        at java.net.DualStackPlainSocketImpl.connect0(Native Method)
        at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:83)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at org.apache.http.conn.socket.PlainConnectionSocketFactory.connectSocket(PlainConnectionSocketFactory.java:74)
        at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:134)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
        at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:55)
        at org.openqa.selenium.remote.internal.ApacheHttpClient.fallBackExecute(ApacheHttpClient.java:144)
        at org.openqa.selenium.remote.internal.ApacheHttpClient.execute(ApacheHttpClient.java:90)
        at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:142)
        at org.openqa.selenium.firefox.internal.NewProfileExtensionConnection.execute(NewProfileExtensionConnection.java:160)
        at org.openqa.selenium.firefox.FirefoxDriver$LazyCommandExecutor.execute(FirefoxDriver.java:380)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:644)
        at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:701)
        at org.openqa.selenium.remote.RemoteWebDriver.quit(RemoteWebDriver.java:526)
        at sun.reflect.GeneratedMethodAccessor23.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.openqa.selenium.support.events.EventFiringWebDriver$2.invoke(EventFiringWebDriver.java:103)
        at com.sun.proxy.$Proxy30.quit(Unknown Source)
        at org.openqa.selenium.support.events.EventFiringWebDriver.quit(EventFiringWebDriver.java:202)
        at org.jenkinsci.test.acceptance.FallbackConfig$1.evaluate(FallbackConfig.java:168)
        at org.jenkinsci.test.acceptance.guice.Cleaner.performCleanUp(Cleaner.java:54)
        at org.jenkinsci.test.acceptance.guice.TestCleaner.performCleanUp(TestCleaner.java:19)
        at org.jenkinsci.test.acceptance.guice.World.endTestScope(World.java:66)
        at org.jenkinsci.test.acceptance.junit.JenkinsAcceptanceTestRule$1.evaluate(JenkinsAcceptanceTestRule.java:72)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
```

### CODE REVIEW

File: src/main/java/org/jenkinsci/test/acceptance/po/SlavesMixIn.java [hunk 1]
```diff
@@ -42,13 +44,14 @@ public SlavesMixIn(Jenkins jenkins) {
     }
 
     private String remoteFs(String name) {
-        String base = "/tmp/";
+        String base = System.getProperty("java.io.tmpdir");
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/po/SlavesMixIn.java]: this is likely to kill the docker ssh slaves test when run on windows...

**Comment by kwhetstone** [comment]
:+1: : x10000000000

**Comment by jtnord** [comment]
ATH is being run against abitrary versions of Jenkins core including PRs and jenkins 2.0 branch (which looks horribly broken to me).  So ignoring any results from [https://jenkins.ci.cloudbees.com/job/core/job/acceptance-test-harness/](https://jenkins.ci.cloudbees.com/job/core/job/acceptance-test-harness/) and only paying attention to [https://jenkins.ci.cloudbees.com/job/core/job/acceptance-test-harness-stable/](https://jenkins.ci.cloudbees.com/job/core/job/acceptance-test-harness-stable/) until the mess is resolved.

File: src/test/java/plugins/PrioritySorterPluginTest.java [hunk 2]
```diff
@@ -34,6 +36,7 @@
 
     @Before
     public void setUp() throws Exception {
+        assumeTrue("This test requires a restartable Jenkins", jenkins.canRestart());
```
Comment (jtnord) [inline src/test/java/plugins/PrioritySorterPluginTest.java]: this would ideally be a rule so that we can avoid the overhead of starting Jenkins for a test that is doomed to failure - but how do we know if we can restart ahead of time :-(

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java [hunk 3]
```diff
@@ -156,6 +156,16 @@ private void copyDockerfileDirectoryFromClasspath(Class<? extends DockerContaine
             // Dockerfile is not packaged into a jar file, so copy locally
             copyDockerfileDirectoryFromLocal(dockerfileLocation, dir);
         }
+        // if the fixture is dynamic (needs to know something about our environment then process it.
+        if (DynamicDockerContainer.class.isAssignableFrom(fixture)) {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: this didn't really work as expected - need to investigate.

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java [hunk 4]
```diff
@@ -16,16 +19,26 @@
 
     public static final String DEFAULT_DOCKER_HOST = "127.0.0.1";
     public final String tag;
-    DockerHostResolver dockerHostResolver;
+    static DockerHostResolver dockerHostResolver = new DockerHostResolver();
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java]: really belongs in `Docker` not `DockerImage`

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java [hunk 5]
```diff
@@ -112,13 +126,16 @@ public DockerImage(String tag) {
         }
     }
 
-    // package for tests
-    String getDockerHost() {
+    /**
+     * Get the string representation of the docker host as set by DOCKER_HOST environment variable or the localhost address if it is not set (or is a socket).
+     * @return an IP Address or hostname of the docker host 
+     */
+    public static String getDockerHost() {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java]: should really be in `Docker` not `DockerImage` - but was already here :-/

File: src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/FtpdContainer.java [hunk 6]
```diff
@@ -18,7 +20,7 @@
  * @author Tobias Meyer
  */
 @DockerFixture(id = "ftpd", ports = {21, 7050, 7051, 7052, 7053, 7054, 7055}, bindIp = "127.0.0.2")
-public class FtpdContainer extends DockerContainer implements IPasswordDockerContainer {
+public class FtpdContainer extends DynamicDockerContainer implements IPasswordDockerContainer {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/FtpdContainer.java]: revert this

File: src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/FtpdContainer.java [hunk 7]
```diff
@@ -58,7 +60,7 @@ public void ftpDisconnect() {
      */
     public Boolean ftpConnect() {
         try {
-            ftpClient.connect(ipBound(21), port(21));
+            ftpClient.connect(getIpAddress(), port(21));
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/FtpdContainer.java]: revert this

File: src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer.java [hunk 8]
```diff
@@ -18,7 +18,7 @@
  *
  * @author Kohsuke Kawaguchi
  */
-@DockerFixture(id = "sshd", ports = 22, bindIp = "127.0.0.5")
+@DockerFixture(id = "sshd", ports = 22)
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer.java]: there was no requirement that I could see for this to be bound to a specific IP address...

File: src/main/java/org/jenkinsci/test/acceptance/junit/WithPlugins.java [hunk 9]
```diff
@@ -91,6 +93,7 @@ public void evaluate() throws Throwable {
                     restartRequired |= installPlugins(d.getTestClass().getAnnotation(WithPlugins.class), plugins);
                     LOGGER.info("for " + d + " asked to install " + plugins + "; restartRequired? " + restartRequired);
                     if (restartRequired) {
+                        assumeTrue("This test requires a restartable Jenkins", jenkins.canRestart());
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/junit/WithPlugins.java]: would be nice to know this ahead of time - but how do you find this out before Jenkins starts :-)

File: src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/FtpdContainer/Dockerfile [hunk 10]
```diff
@@ -20,6 +20,10 @@ RUN echo "rsa_cert_file=/etc/ssl/private/vsftpd.pem" >> /etc/myftp.conf
 RUN echo "pasv_enable=YES" >> /etc/myftp.conf
 RUN echo "pasv_min_port=7050" >> /etc/myftp.conf
 RUN echo "pasv_max_port=7055" >> /etc/myftp.conf
+# To enable passv mode through NAT (for remote docker) we need to set the 
+# passv_address to the address that docker will use to port forward
+RUN echo "pasv_address=@@DOCKER_HOST@@" >> /etc/myftpd.conf
```
Comment (jtnord) [inline src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/FtpdContainer/Dockerfile]: revert this (and investigate why this was not set!)

**Comment by jtnord** [comment]
@reviewbybees /cc @olivergondza  Still not ready for merge but I would appreciate any feedback on the direction this is going...

**Comment by olivergondza** [comment]
So the way docker is used on windows resembles non-local docker host, right?

**Comment by jtnord** [comment]
> So the way docker is used on windows resembles non-local docker host, right?

@olivergondza yes - it's effectively the same as running on Linux with `DOCKER_HOST=tcp://some.remote.machine:port/`  (which is what you need to do on Mac as well)

**Comment by jtnord** [comment]
> I seem to have broken docker portmap (Failed to figure out port map 80)

Or not.  I can one of the failing tests with `mvn test -Dtest=XvncPluginTest` and it passes
`Tests run: 4, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 391.441 sec - in plugins.XvncPluginTest`

Which implies something environmental or an issue with docker :-(

From the logs
> docker: Error response from daemon: failed to create endpoint nauseous_goodall on network bridge: Error starting userland proxy: listen tcp 192.168.99.100:32787: bind: cannot assign requested address.

File: src/test/java/plugins/CompressArtifactsPluginTest.java [hunk 11]
```diff
@@ -101,11 +119,6 @@ public void archiveLargerThan4GInTotal() throws Exception {
                 job.name
         ));
         assertThat(length, greaterThanOrEqualTo(4L * 1024 * 1024 * 1024));
-
```
Comment (jtnord) [inline src/test/java/plugins/CompressArtifactsPluginTest.java]: this code just consumed cycles and did not test anything.

File: src/test/java/plugins/CompressArtifactsPluginTest.java [hunk 12]
```diff
@@ -101,11 +119,6 @@ public void archiveLargerThan4GInTotal() throws Exception {
...
-        build.getArtifact("stuff.jar").open();
-        for (int i = 1; i <= 70; i++) {
-            build.getArtifact("stuff." + i + "42.jar").open();
```
Comment (jtnord) [inline src/test/java/plugins/CompressArtifactsPluginTest.java]: would attempt to open `stuff.7042.jar` which is not created and would not exist yet the test would still pass anyway.

File: src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java (hunk 3 above)
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/docker/Docker.java]: Works as expected - but the portmapping was still getting in the way for the FTP PASSV connections.

File: src/main/java/org/jenkinsci/test/acceptance/junit/Native.java [hunk 13]
```diff
@@ -46,8 +47,15 @@ public void evaluate() throws Throwable {
                 private void verifyNativeCommandPresent(Native n) throws IOException, InterruptedException {
                     if (n==null)        return;
                     for (String cmd : n.value()) {
-                        if (new CommandBuilder("which",cmd).system()!=0) {
-                            throw new AssumptionViolatedException(cmd + " is needed for the test but doesn't exist in the system");
+                        if (SystemUtils.IS_OS_WINDOWS) {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/junit/Native.java]: this tests the command is available locally - not on the machine running Jenkins which may be different.  Hey ho - this was previously the case and is likely to be the case most of the time.  A larger fix will require starting Jenkins which is the whole point of this annotation to begine with (avoids starting Jenkins if the command is not present)

File: src/main/java/org/jenkinsci/test/acceptance/junit/WithPlugins.java (hunk 9 above)
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/junit/WithPlugins.java]: future possibility could be to fallback to the `JenkinsController` and ask it to perform a restart.

File: src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java [hunk 14]
```diff
@@ -42,20 +43,45 @@ public DumbSlave(Jenkins j, String name) {
...
      */
     public void asLocal() {
-        assertCurl();
-        File jar = new File("/tmp/slave"+createRandomName()+".jar");
-        String command = String.format(
-                "sh -c 'curl -s -o %1$s %2$sjnlpJars/slave.jar && java -jar %1$s'",
-                jar, url("../../")
-        );
+        assertCurlOrPowershellv3Plus();
+        File tmpDir = new File(System.getProperty("java.io.tmpdir"));
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java]: Not convinced this should actually use tmp, but rather the build directory so mvn clean gets rid of the garbage but whatever.

File: src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java [hunk 15]
```diff
@@ -42,20 +43,45 @@ public DumbSlave(Jenkins j, String name) {
...
+            command = String.format(
+                    "sh -c 'curl -s -o %1$s %2$sjnlpJars/slave.jar && java -jar %1$s'",
+                    jar, url("../../")
+            );
+        }
+        else {
+            // windows
+            command = String.format("powershell -command \"& { " +
+                                                               "try { " +
+                                                                  "Invoke-WebRequest %2$sjnlpJars/slave.jar -OutFile %1$s " +
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/po/DumbSlave.java]: finally a use for powershell :-)

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java [hunk 16]
```diff
@@ -16,16 +19,26 @@
...
         this.tag = tag;
-        dockerHostResolver = new DockerHostResolver();
     }
 
-    public <T extends DockerContainer> T start(Class<T> type, CommandBuilder options, CommandBuilder cmd,int portOffset) throws InterruptedException, IOException {
+    public <T extends DockerContainer> T start(Class<T> type, CommandBuilder options, CommandBuilder cmd, int portOffset) throws InterruptedException, IOException {
         DockerFixture f = type.getAnnotation(DockerFixture.class);
-        return start(type,f.ports(),portOffset,f.bindIp(),options,cmd);
+
+        String addr = f.bindIp();
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java]: Didn't you deprecate `bindIp`? And wouldn't that mean that you are no longer using it?

Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java]: I deprecated it to say "Hey you better have a good reason for using this"  as this was the only legitimate use that I found - and you also need to add `withLocal=true`

If I re-wrote the `FTPPublisherTest` then we could completely remove it completely and run this test on windows - but I felt that that was a little bit of scope creep :-)

File: src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/SvnContainer.java [hunk 17]
```diff
@@ -45,7 +44,7 @@ public URL getHttpUrl() throws SubversionPluginTestException {
      * @throws SubversionPluginTestException e
      */
     public URI getSvnUrl() throws SubversionPluginTestException {
-        String url = PROTOCOL_SVN + LOCALHOST + port(3690);
+        String url = PROTOCOL_SVN + ipBound(80) + ':' + port(3690);
```
Comment (mslusarczyk) [inline src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/SvnContainer.java]: is it not supposed to be ipBound(3690)?

Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/docker/fixtures/SvnContainer.java]: it could do - but both the ssh daemon and the web interface will be on the same machine so....

**Comment by jtnord** [comment]
#630198d should fix the failing test introduced earlier (due to a bug in selenioum).

Following failures are not new and can be ignored.
```
plugins.PrioritySorterPluginTest.match_jobs_by_view
plugins.NodeLabelParameterPluginTest.run_on_online_slave_and_master_with_node_restriction
plugins.NodeLabelParameterPluginTest.run_on_several_slaves
plugins.AntPluginTest.locallyInstalledAnt
```
`NodeLabelParameterPluginTest` legitimate failure filed as [JENKINS-33746](https://issues.jenkins-ci.org/browse/JENKINS-33746)
`plugins.PrioritySorterPluginTest.match_jobs_by_view` seems to be caused by async loading of runs?
`plugins.AntPluginTest.locallyInstalledAnt` not investigated - but failing on the master branch.

File: pom.xml [hunk 18]
```diff
@@ -160,8 +160,8 @@
     </dependency>
     <dependency>
       <groupId>org.seleniumhq.selenium</groupId>
-      <artifactId>selenium-htmlunit-driver</artifactId>
-      <version>${selenium.version}</version>
+      <artifactId>htmlunit-driver</artifactId>
+      <version>2.20</version>
```
Comment (jtnord) [inline pom.xml]: the htmlunit-driver  was [split out](https://github.com/SeleniumHQ/htmlunit-driver) of the main selenium project and has it's own lifecycle.

File: src/main/java/org/jenkinsci/test/acceptance/junit/WithPlugins.java (hunk 9 above)
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/junit/WithPlugins.java]: I have a follow on PR that does exactly this - but I want to get this merged first.

**Comment by kzantow** [comment]
I'm sure @tfennelly wants to review this one :-D

**Comment by mslusarczyk** [comment]
LGTM :bee:

**Comment by andresrc** [comment]
:bee:

File: src/test/java/plugins/LdapPluginTest.java [hunk 19]
```diff
@@ -205,7 +210,7 @@ public void login_use_fallback_server() {
         int freePort = this.findAvailablePort();
         LdapDetails ldapDetails = new LdapDetails("", 0, ldapContainer.getManagerDn(), ldapContainer.getManagerPassword(), ldapContainer.getRootDn());
         // Fallback-Config: primary server is not running, alternative server is running docker fixture
-        ldapDetails.setHostWithPort("localhost:" + freePort + " localhost:" + ldapContainer.getPort());
+        ldapDetails.setHostWithPort("localhost:" + freePort + ' ' + ldapContainer.getHost() + ':' + ldapContainer.getPort());
```
Comment (armfergom) [inline src/test/java/plugins/LdapPluginTest.java]: Same change in this [PR](https://github.com/jenkinsci/acceptance-test-harness/pull/82)

File: src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java [hunk 20]
```diff
@@ -115,6 +116,16 @@ public PluginManager getPluginManager() {
         return new PluginManager(this);
     }
 
+    /** 
+     * Some tests require they restart Jenkins - but depending on how the SUT is launched this is not always possible
+     * so tests that require this should wrap this call in an {@link org.junit.Assume#assumeTrue(String, boolean)}
+     * @return true if the Jenkins under test can restart itself.
+     */
+    public boolean canRestart() {
```
Comment (armfergom) [inline src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java]: May I ask under which circumstances a Jenkins instance is not restartable?

**Comment by armfergom** [comment]
:bee:

File: src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java (hunk 20 above)
Comment (jtnord) [inline src/main/java/org/jenkinsci/test/acceptance/po/Jenkins.java]: By default on windows when run with `java -jar` is the big one :)

**Comment by jtnord** [comment]
@reviewbybees done

File: src/test/java/plugins/CompressArtifactsPluginTest.java [hunk 21]
```diff
@@ -89,9 +91,25 @@ public void archiveLargerThan4GInTotal() throws Exception {
...
                 "#!/bin/bash\nwget $JENKINS_URL/jnlpJars/jenkins-cli.jar -O stuff.jar; for i in {0..7000}; do cp -l stuff.jar stuff.${i}.jar; done"
-        );
+                                            );
+        }
+        else {
+            // On windows although we could create hard links this would exceed the number of hard links you can create for a given file. 
+            // and sparse files compress too well. 
+            //job.addBuildStep(BatchCommandBuildStep.class).command("for /l %%x in (0, 1, 7000) do fsutil file createnew stuff.%%x.jar 819200");
+            job.addBuildStep(BatchCommandBuildStep.class).command("powershell -command \"& { " +
+                                            "try { " +
```
Comment (rsandell) [inline src/test/java/plugins/CompressArtifactsPluginTest.java]: I guess this should have a similar `assumeWgetOrPowerShellv3` thingie as you've done in other places?
But since the test didn't have that before I don't see it as any blocker, just a nice to have.

**Comment by rsandell** [comment]
:bee: on the last commit

File: src/test/java/plugins/CompressArtifactsPluginTest.java (hunk 21 above)
Comment (jtnord) [inline src/test/java/plugins/CompressArtifactsPluginTest.java]: According to [this page](https://4sysops.com/archives/powershell-versions-and-their-windows-version/) the version of PowerShell on supported windows platforms (8+ or 2012+) is at least 3 and IIUC is installed by default - so the check would be redundant unless we are running on an unsupported platform at which point meh!

**Comment by jtnord** [comment]
The new test failures in the last run exist in other PR builds - so not introduced with this change.

**Comment by jtnord** [comment]
@rsandell @olivergondza I'm calling this done and ready for merge.
